src/pepr/operator/controllers/keycloak/clients/client-credentials.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright 2025 Defense Unicorns
+ * SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Defense-Unicorns-Commercial
+ */
+
+import { fetch, K8s, kind } from "pepr";
+import { Client } from "../types";
+import { baseUrl, log, throwErrorIfNeeded } from "./common";
+
+export interface ClientWithId extends Client {
+  id: string;
+}
+
+export interface KeycloakAccessTokenResponse {
+  access_token: string;
+}
+
+export function parseKeycloakToken(token: string) {
+  return JSON.parse(Buffer.from(token.split(".")[1], "base64").toString()) as KeycloakToken;
+}
+
+const clientsAdminUrl = `${baseUrl}/admin/realms/uds/clients`;
+const clientCredentialsUrl = `${baseUrl}/realms/uds/protocol/openid-connect/token`;
+const SECRET_NAMESPACE = "keycloak";
+const SECRET_NAME = "keycloak-client-secrets";
+const UDS_OPERATOR_CLIENT_ID = "uds-operator";
+let cachedToken: string | null = null;
+
+export interface KeycloakToken {
+  exp: number;
+  resource_access: {
+    "realm-management": {
+      roles: string[];
+    };
+    [key: string]: unknown;
+  };
+
+  [key: string]: unknown;
+}
+
+export function resetCachedToken() {
+  cachedToken = null;
+}
+
+export async function credentialsGetAccessToken() {
+  if (cachedToken) {
+    try {
+      const jwt = parseKeycloakToken(cachedToken);
+      if (jwt.exp && jwt.exp > Math.floor(Date.now() / 1000) + 5) return cachedToken;
+    } catch (e) {
+      log.error(e, "Failed to parse cached token");
+      cachedToken = null;
+    }
+  }
+
+  const secret = await K8s(kind.Secret).InNamespace(SECRET_NAMESPACE).Get(SECRET_NAME);
```

Comment:
I'm not totally against this, but I wonder if it makes sense to just cache this secret? Since we have a watch on the secret, we can "refresh the cache" inside the `updateKeycloakClientsSecret` function and avoid needing the network call here?